### ISSUE TITLE
Problems matching more complicated laws

### ISSUE THREAD

**Issue by sjcjoosten** [nicolasjulioflores/calculus-reasoner#2]
Currently, all laws are of the form d/dx (Opr f [...]) where the ... only contains variables.
In this case, mustMatch makes sense, but its final cases, including ``mustMatch _ _``, are problematic as it pretends to succesfully match something. It might be a better idea to turn mustMatch into something that returns a list of potential substitutions, so it can return the empty list for this final case.

In your function match', you are checking that your final substitution is not the empty substitution, which takes care of a lot of the problems for you, but this is problematic in itself: you now are unable to match a law like "-0.0 = 0.0"

**Comment by nicolasjulioflores** [comment]
I've included an example of the "-0.0 = 0.0" law in our most recent commit and you can test it in ghci with " map (calculate zlaws) [ex_z, ex_v]", with the relevant code in Lib.hs. I think we've had to make our types quite ugly to get it to work though.

**Comment by sjcjoosten** [comment]
I think the underlying problems is that the empty substitution should not be treated the same as no substitution. Why not change the type of functions like ``combine`` to return ``[Subst]`` or ``Maybe Subst`` instead of ``Subst``?

**Comment by nicolasjulioflores** [comment]
Yes that makes sense, I'll add a TODO to change the type to a Maybe Subst so that we can do a less hacky check to see if the substitutions are being made.

Do you have any comments on how to improve our Equation / Transform type (Lib.hs)? and any comments on how our overall reasoning are looking?

**Comment by sjcjoosten** [comment]
To answer your cleanup question, your `Transform` does not belong into an expression because it is not an expression. As you only use it on the right hand side of a law, that should be easy to do. It means that you might want to change your law type and give that a couple of cases. It can be either an equation, or, to be perfectly general, something that does a rewrite step for you without matching, so a function ``Expr -> Maybe Expr`` for instance. Alternatively, you could add ``(Expr, Transform)`` as a case in ``Law`` or ``CLaw``.

I'd also move common parts (like the lawname) into ``CLaw`` and have ``Law`` be one of its cases, so everything put together could be something like this:

```
data CLaw = CLaw Lawname Expr Replacement -- name, the expression to match on, and the replacement to be made
data Replacement = Conditional Condition Replacement
     | WithExpr Expr
     | BuiltIn ([(Variable, Expr)] -> Maybe Expr)
```

If you feel confident using very general abstract types, consider using this (no ``CLaw`` needed):

```
data Law = Law Lawname (Expr -> Maybe Expr)
```

Printing these laws will become impossible, and while parsing 'normal' laws you'll have to use functions like ``match``.
